sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stddef.h>
+
+#include "sw/device/lib/dif/dif_padctrl.h"
+
+#include "padctrl_regs.h"
+#include "sw/device/lib/base/mmio.h"
+
+#define PADCTRL_REGISTER_WIDTH_IN_BYTES (PADCTRL_PARAM_REG_WIDTH / 8)
+
+// This is the value written to a whole attribute register, it is not a per-pad
+// value so desired default attribute setting must be replicated per attribute
+// field in the attribute register.
+#define PADCTRL_ATTR_DEFAULT 0
+#define PADCTRL_ATTR_FIELD_MASK PADCTRL_MIO_PADS0_ATTR0_MASK
+
+/**
+ * Checks if an attribute field contains conflicting attributes
+ */
+static bool dif_padctrl_conflicting_attributes(uint32_t attrs) {
+  // Cannot be both pull up and pull down
+  if ((attrs & (1 << kDifPadctrlAttrPullDown)) &&
+      (attrs & (1 << kDifPadctrlAttrPullUp))) {
+    return true;
+  }
+
+  // Cannot combine pull down with open drain
+  if ((attrs & (1 << kDifPadctrlAttrPullDown)) &&
+      (attrs & (1 << kDifPadctrlAttrOpenDrain))) {
+    return true;
+  }
+
+  return false;
+}
+
+dif_padctrl_init_error_t dif_padctrl_init(mmio_region_t base_addr,
+                                          dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlInitErrorInvalidPadctrl;
+  }
+
+  bool padctrl_unlocked = mmio_region_get_bit32(
+      base_addr, PADCTRL_REGEN_REG_OFFSET, PADCTRL_REGEN_WEN);
+
+  if (!padctrl_unlocked) {
+    return kDifPadctrlInitErrorRegistersLocked;
+  }
+
+  padctrl->base_addr = base_addr;
+
+  // Reset all DIO pad attributes to the default value
+  for (int i = 0; i < PADCTRL_DIO_PADS_MULTIREG_COUNT; ++i) {
+    // TODO: Fix reggen so DIO_PADS defines don't switch between having and not
+    // having a 0 depending on whether multiple regs are needed.
+    ptrdiff_t dio_pads_register_offset =
+        PADCTRL_DIO_PADS_REG_OFFSET + (i * PADCTRL_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(padctrl->base_addr, dio_pads_register_offset,
+                        PADCTRL_ATTR_DEFAULT);
+  }
+
+  // Reset all MIO pad attributes to the default value
+  for (int i = 0; i < PADCTRL_MIO_PADS_MULTIREG_COUNT; ++i) {
+    // TODO: Fix reggen so MIO_PADS defines don't switch between having and not
+    // having a 0 depending on whether multiple regs are needed.
+    ptrdiff_t mio_pads_register_offset =
+        PADCTRL_MIO_PADS0_REG_OFFSET + (i * PADCTRL_REGISTER_WIDTH_IN_BYTES);
+    mmio_region_write32(padctrl->base_addr, mio_pads_register_offset,
+                        PADCTRL_ATTR_DEFAULT);
+  }
+
+  return kDifPadctrlInitErrorNone;
+}
+
+dif_padctrl_registers_lock_error_t dif_padctrl_registers_lock(
+    dif_padctrl_t *padctrl) {
+  if (padctrl == NULL) {
+    return kDifPadctrlRegistersLockErrorInvalidPadctrl;
+  }
+
+  mmio_region_write32(padctrl->base_addr, PADCTRL_REGEN_REG_OFFSET, 0);
+
+  return kDifPadctrlRegistersLockErrorNone;
+}
+
+dif_padctrl_enable_attr_error_t dif_padctrl_enable_attr(
```

Comment:
Shouldn't this function check if padctrl is locked and shouldn't it be another error condition?

Reply:
I did have a bit of discussion with @lenary and @silvestrst about this point. My general principle was you do a single lock check at init time then leave it up to the driving software to not call into the DIF once its locked it (or see it was locked at init time).

I could add a lock check here but it would mean an extra register read that the vast majority of the time is pointless.